src/ape/utils/abi.py

```diff
@@ -285,45 +285,47 @@ def decode_value(self, abi_type, value):
 
         return value
 
+def parse_type(type: Dict[str, Any]) -> Union[str, Tuple, List]:
+    """Parses ABIType.dict() into Python types.
+
+    >>> parse_type({'type': 'uint256'})
+    'uint256'
+    >>> parse_type({'type': 'tuple', 'components': [{'type': 'uint256'}]})
+    ('uint256',)
+    >>> parse_type({'type': 'tuple', 'components': [{'type': 'uint256'}, {'type':\
+    'uint256'}]})
+    ('uint256', 'uint256')
+    >>> parse_type({'type': 'tuple', 'components': [{'type': 'uint256'}, {'type':\
+    'tuple', 'components': [{'type': 'uint256'}]}]})
+    ('uint256', ('uint256',))
+    >>> parse_type({'type': 'tuple[]', 'components': [{'type': 'uint256'}]})
+    [('uint256',)]
+    >>> parse_type({'type': 'tuple[]', 'components': [{'type': 'uint256'}, {'type':\
+    'tuple', 'components': [{'type': 'uint256[]'}]}]}) 
+    [('uint256', (['uint256'],))]
+    >>> parse_type({'type': 'tuple', 'components': [{'type': 'tuple[]',\
+    'components': [{'type': 'int[]'}]}, {'type': 'int'}]})
+    ([(['int'],)], 'int')
+    >>> parse_type({'type': 'int[][]'})
+    [['int']]
 
-def parse_type(output_type: str) -> Union[str, Tuple, List]:
-    if not output_type.startswith("("):
-        return output_type
-
-    # Strip off first opening parens
-    output_type = output_type[1:]
-    found_types: List[Union[str, Tuple, List]] = []
-
-    while output_type:
-        if output_type.startswith(")"):
-            result = tuple(found_types)
-            if "[" in output_type:
-                return [result]
-
-            return result
-
-        elif output_type[0] == "(" and ")" in output_type:
-            # A tuple within the tuple
-            end_index = output_type.index(")") + 1
-            found_type = parse_type(output_type[:end_index])
-            output_type = output_type[end_index:]
-
-            if output_type.startswith("[") and "]" in output_type:
-                end_array_index = output_type.index("]") + 1
-                found_type = [found_type]
-                output_type = output_type[end_array_index:].lstrip(",")
-
-        else:
-            found_type = output_type.split(",")[0].rstrip(")")
-            end_index = len(found_type) + 1
-            output_type = output_type[end_index:]
-
-        if isinstance(found_type, str) and "[" in found_type and ")" in found_type:
-            parts = found_type.split(")")
-            found_type = parts[0]
-            output_type = f"){parts[1]}"
-
-        if found_type:
-            found_types.append(found_type)
+    Args:
+        type (Dict[str, Any]): This is ABIType.dict(). The two keys
+        that we are interested in are `type` and `components`.
+        `components` produces the same two keys (if present),
+        namely `type` and `components`.
 
-    return tuple(found_types)
+    Returns:
+        Union[str, Tuple, List]: Python representation of the type.
+        For example, `uint256[]` becomes ['uint256'].
+    """
+    if type["type"] == "tuple[]":
```

Comment:
I get tests to pass when this also handles fixed length arrays like `tuple[2]` etc.
Maybe we can swap this check with a regex-like check instead?

`decode_primitive_value()` expects these solidity types to have the component-canonical types in the type-strings but they are missing by the time it gets to the part where it decodes the values. That is how the type + value relate in the context of these failing tests.

Reply:
What's the definition of `component-canonical types`?

Reply:
what should parse type produce when we get `uint256[2][3]`?

```python
def parse_type(type):
    match = re.search(r"(^.+?(?=\[|$))(.*)", type["type"])
    if match is None:
        raise ValueError(f"Invalid type: {type['type']}")
    primitive_type, nesting = match.groups()
    num_brackets = nesting.count("[")

    if primitive_type == "tuple" and not nesting:
        return tuple([parse_type(c) for c in type["components"]])
    elif primitive_type == "tuple" and nesting:
        result = tuple([parse_type(c) for c in type["components"]])
        for _ in range(num_brackets):
            result = [result]
        return result
    if nesting:
        result = primitive_type
        for _ in range(num_brackets):
            result = [result]
        return result
    else:
        return primitive_type 
```

returns `[['uint256']]`, but then somewhere down the line this is turned into tuples and the actual output has tuples in there, and the expected one does not.

Reply:
going to answer this myself. Just ran doctests on the current main, and in fact for `uint256[2][3]` it just returns that.

```
Failed example:
    parse_type('uint256[2][3]')
Expected:
    [['uint256']]
Got:
    'uint256[2][3]'
```

Reply:
this is very weird, because a `uint256[][]` would be `[['uint256']]`